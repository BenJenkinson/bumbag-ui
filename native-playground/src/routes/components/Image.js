import React from 'react';
import { Box, Image, Text } from 'bumbag-native';
import { Preview } from '../../components/Preview';
import { PreviewSection } from '../../components/PreviewSection';

export default function App() {
  return (
    <Box>
      <Text fontSize="20px" fontWeight="500" marginBottom="16px">
        Image
      </Text>
      <PreviewSection title="Basic">
        <Preview>
          <Box>
            <Image source={require('../../images/bean.jpg')} width="500px" height="300px" />
          </Box>
        </Preview>
      </PreviewSection>
      <PreviewSection title="Fit - Contain">
        <Preview>
          <Box>
            <Image
              backgroundColor="primary"
              source={require('../../images/bean.jpg')}
              width="300px"
              height="300px"
              fit="contain"
            />
          </Box>
        </Preview>
      </PreviewSection>
      <PreviewSection title="Fit - Cover">
        <Preview>
          <Box>
            <Image
              backgroundColor="primary"
              source={require('../../images/bean.jpg')}
              width="300px"
              height="300px"
              fit="stretch"
            />
          </Box>
        </Preview>
      </PreviewSection>
      <PreviewSection title="Fit - Repeat">
        <Preview>
          <Box>
            <Image
              backgroundColor="primary"
              source={require('../../images/bean.jpg')}
              width="300px"
              height="300px"
              fit="repeat"
            />
          </Box>
        </Preview>
      </PreviewSection>
      <PreviewSection title="Fit - Center">
        <Preview>
          <Box>
            <Image
              backgroundColor="primary"
              source={require('../../images/bean.jpg')}
              width="300px"
              height="300px"
              fit="center"
            />
          </Box>
        </Preview>
      </PreviewSection>
    </Box>
  );
}
