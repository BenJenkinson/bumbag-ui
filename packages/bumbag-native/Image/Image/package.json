{
  "name": "bumbag-native/Image/Image",
  "private": true,
  "sideEffects": false,
  "main": "../../lib/Image/Image",
  "module": "../../es/Image/Image",
  "types": "../../ts/Image/Image"
}