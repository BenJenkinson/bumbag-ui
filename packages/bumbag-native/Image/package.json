{
  "name": "bumbag-native/Image",
  "private": true,
  "sideEffects": false,
  "main": "../lib/Image",
  "module": "../es/Image",
  "types": "../ts/Image"
}