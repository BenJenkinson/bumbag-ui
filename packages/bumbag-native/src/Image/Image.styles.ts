import { Image } from 'react-native';
import { styled } from '../styled';
import { theme } from '../utils/theme';

export const StyledImage = styled(Image)`
  ${theme('Image', 'styles.base')};
`;
