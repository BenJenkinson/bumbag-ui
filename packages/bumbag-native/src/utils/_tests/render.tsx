import * as React from 'react';
import { ThemeConfig } from '../../types/theme';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (
  Component: any,
  { colorMode = 'default', theme = {} }: { colorMode?: string; theme?: ThemeConfig } = {}
) =>
  render(
    <Provider platform="native" theme={theme} colorMode={colorMode}>
      {Component}
    </Provider>
  );
