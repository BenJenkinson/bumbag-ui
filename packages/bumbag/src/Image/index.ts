import * as styles from './styles';

export * from './Image';
export { styles as imageStyles };
