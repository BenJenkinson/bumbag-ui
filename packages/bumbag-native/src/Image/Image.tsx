import { ImageProps as RNImageProps } from 'react-native';
import { createComponent, createElement, createHook } from 'bumbag/utils';

import { Box, BoxProps } from '../Box';
import * as styles from './Image.styles';

export type LocalImageProps = {
  /** How the image fits its bounds */
  fit?: RNImageProps['resizeMode'];
};
export type ImageProps = BoxProps & RNImageProps & LocalImageProps;

const useProps = createHook<ImageProps>(
  (props) => {
    const boxProps = Box.useProps(props);
    return { ...boxProps };
  },
  { themeKey: 'native.Image' }
);

export const Image = createComponent<ImageProps>(
  (props) => {
    const imageProps = useProps(props);
    return createElement({
      children: props.children,
      component: styles.StyledImage,
      htmlProps: {
        ...imageProps,
        resizeMode: props.fit || props.resizeMode,
      },
    });
  },
  {
    attach: {
      useProps,
      displayName: 'native.Image',
    },
    themeKey: 'native.Image',
  }
);
