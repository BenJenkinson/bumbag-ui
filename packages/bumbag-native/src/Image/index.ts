import * as imageStyles from './Image.styles';

export * from './Image';
export { imageStyles };
