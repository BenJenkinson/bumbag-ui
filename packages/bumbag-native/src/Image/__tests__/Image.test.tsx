import * as React from 'react';
import { Image } from '../Image';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(<Image accessibilityLabel="bean" source={{ uri: 'test.png' }} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(
      <Image accessibilityLabel="Bean" backgroundColor="red" source={{ uri: 'test.png' }} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly for fit contain', () => {
    const { container } = render(<Image accessibilityLabel="Bean" fit="contain" source={{ uri: 'test.png' }} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('Image.base should render correctly', () => {
    const { container } = render(<Image source={{ uri: 'test.png' }} />, {
      theme: { Image: { styles: { base: { backgroundColor: 'red' } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for default props', () => {
    const { container } = render(<Image source={{ uri: 'test.png' }} />, {
      theme: { Image: { defaultProps: { accessibilityLabel: 'Bean' } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
