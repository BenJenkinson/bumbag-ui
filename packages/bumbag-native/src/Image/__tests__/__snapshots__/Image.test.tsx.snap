// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultProps should render correctly for default props 1`] = `
<div
  aria-label="Bean"
  class="css-view-1dbjc4n"
  style="flex-basis: auto; overflow-x: hidden; overflow-y: hidden; z-index: 0;"
>
  <div
    class="css-view-1dbjc4n"
    style="background-color: rgba(0, 0, 0, 0); background-position: center; background-repeat: no-repeat; background-size: cover; bottom: 0px; height: 100%; left: 0px; position: absolute; right: 0px; top: 0px; width: 100%; z-index: -1; background-image: url(test.png);"
  />
  <img
    alt="Bean"
    class="css-accessibilityImage-9pa8cd"
    draggable="false"
    src="test.png"
  />
</div>
`;

exports[`props should render correctly 1`] = `
<div
  aria-label="bean"
  class="css-view-1dbjc4n"
  style="flex-basis: auto; overflow-x: hidden; overflow-y: hidden; z-index: 0;"
>
  <div
    class="css-view-1dbjc4n"
    style="background-color: rgba(0, 0, 0, 0); background-position: center; background-repeat: no-repeat; background-size: cover; bottom: 0px; height: 100%; left: 0px; position: absolute; right: 0px; top: 0px; width: 100%; z-index: -1; background-image: url(test.png);"
  />
  <img
    alt="bean"
    class="css-accessibilityImage-9pa8cd"
    draggable="false"
    src="test.png"
  />
</div>
`;

exports[`props should render correctly for fit contain 1`] = `
<div
  aria-label="Bean"
  class="css-view-1dbjc4n"
  style="flex-basis: auto; overflow-x: hidden; overflow-y: hidden; z-index: 0;"
>
  <div
    class="css-view-1dbjc4n"
    style="background-color: rgba(0, 0, 0, 0); background-position: center; background-repeat: no-repeat; background-size: contain; bottom: 0px; height: 100%; left: 0px; position: absolute; right: 0px; top: 0px; width: 100%; z-index: -1; background-image: url(test.png);"
  />
  <img
    alt="Bean"
    class="css-accessibilityImage-9pa8cd"
    draggable="false"
    src="test.png"
  />
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
<div
  aria-label="Bean"
  class="css-view-1dbjc4n"
  style="background-color: rgb(255, 0, 0); flex-basis: auto; overflow-x: hidden; overflow-y: hidden; z-index: 0;"
>
  <div
    class="css-view-1dbjc4n"
    style="background-color: rgba(0, 0, 0, 0); background-position: center; background-repeat: no-repeat; background-size: cover; bottom: 0px; height: 100%; left: 0px; position: absolute; right: 0px; top: 0px; width: 100%; z-index: -1; background-image: url(test.png);"
  />
  <img
    alt="Bean"
    class="css-accessibilityImage-9pa8cd"
    draggable="false"
    src="test.png"
  />
</div>
`;

exports[`theming Image.base should render correctly 1`] = `
<div
  class="css-view-1dbjc4n"
  style="background-color: rgb(255, 0, 0); flex-basis: auto; overflow-x: hidden; overflow-y: hidden; z-index: 0;"
>
  <div
    class="css-view-1dbjc4n"
    style="background-color: rgba(0, 0, 0, 0); background-position: center; background-repeat: no-repeat; background-size: cover; bottom: 0px; height: 100%; left: 0px; position: absolute; right: 0px; top: 0px; width: 100%; z-index: -1; background-image: url(test.png);"
  />
  <img
    alt=""
    class="css-accessibilityImage-9pa8cd"
    draggable="false"
    src="test.png"
  />
</div>
`;
